# Causal Inference Using Pearl's Framework

## Objectives:

- Perform a causal inference task using Pearl’s framework;
- Infer the causal graph from observational data and then validate the graph;
- Merge machine learning with causal inference;


## Load Data and Libraries

In [14]:
# Libraries

import pandas as pd
import numpy as np

In [4]:
# Load Data

raw_df = pd.read_csv("../data/data.csv")
raw_df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [5]:
# Check Dataset

raw_df.shape

(569, 33)

In [6]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

### Observation
- There are 569 rows and 33 columns
- The last column is completely empty
- There is no Null value in any of the other columns.
- All the variables are float type except for the Id column and the diagnosis variable which is string.

## Clean Data

In [7]:
# removing null column
clean_df = raw_df.iloc[:,:-1]
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

### check for outliers

In [22]:
# Use IQR to create outlier checking range

tmp_info = clean_df.iloc[:,1:].describe()

Q1 = np.array(tmp_info.iloc[4,:].values.flatten().tolist())
Q3 = np.array(tmp_info.iloc[6,:].values.flatten().tolist())

# calculate the Inerquartile range.
IQR = Q3-Q1
L_factor = IQR*1.5
H_factor = IQR*3

# Minor Outliers will lie outside the Inner fence
Inner_Low = Q1-L_factor
Inner_High = Q3 + L_factor

# Major Outliers will lie outside the Outer fence
Outer_Low = Q1-H_factor
Outer_High = Q3+H_factor


In [ ]:
clean_df["minor_outlier"] = clean_df



In [10]:
tmp_info.iloc[[4,6],]

id  radius_mean  texture_mean  perimeter_mean  area_mean  \
25%   869218.0        11.70         16.17           75.17      420.3   
75%  8813129.0        15.78         21.80          104.10      782.7   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
25%          0.08637           0.06492         0.02956              0.02031   
75%          0.10530           0.13040         0.13070              0.07400   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
25%         0.1619  ...         13.01          21.08            84.11   
75%         0.1957  ...         18.79          29.72           125.40   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
25%       515.3            0.1166             0.1472           0.1145   
75%      1084.0            0.1460             0.3391           0.3829   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
25%               0.06493          0.2504                  0.07146  
75%               0.16140          0.3179                  0.09208  

[2 rows x 31 columns]

## Extract Features

## Perform Exploratory Analysis

## Discover Causal Graphs

## Evaluate Causal Graphs and Optimize

## Model

## Compare Models